tests/main/tests/integration/record-array-test.js

```diff
@@ -400,4 +400,67 @@ module('unit/record-array - RecordArray', function (hooks) {
 
     assert.true(people.isLoaded, 'The array is now loaded');
   });
+
+  test('a record array should only remove object(s) if found in collection', async function (assert) {
+    store.push({
+      data: [
+        {
+          type: 'person',
+          id: '1',
+          attributes: {
+            name: 'Scumbag Dale',
+          },
+        },
+        {
+          type: 'person',
+          id: '2',
+          attributes: {
+            name: 'Scumbag Katz',
+          },
+        },
+        {
+          type: 'person',
+          id: '3',
+          attributes: {
+            name: 'Scumbag Bryn',
+          },
+        },
+      ],
+    });
+
+    const personNotInData = {
+      type: 'person',
+      id: '5',
+      attributes: {
+        name: 'Scumbag Ross',
+      },
+    };
+
+    let recordArray = store.peekAll('person');
```

Comment:
EmberData has 3 kinds of managed record arrays (all of which now utilize `IdentifierArray`, the reason this test fails is because you chose one of the "immutable" ones :D I'll explain

`peekAll` (and by extension `findAll` which returns peekAll once it resolves) return what we refer to as a `LiveArray`, this is a managed record-array that reflects all records loaded for a given type in the store. Since this reflects cache state but cannot mutate cache state, we error if mutation methods are called (this is why this test fails).

`query` returns a managed record array containing a subset of records specific to that query. Historically we referred to this as an `AdapterPopulatedRecordArray` though that is no longer a great name. The distinction now is merely over whether the array reflects a specific collection or "all". In the codebase this is now referred to as a `Collection`. Currently you only gain access to one of these via a `query` but this is about to change (emberjs/rfcs#854 and emberjs/rfcs#860 collectively specify the mechanics of more generalized collections).

`hasMany` relationships are the final managed array (historically called a `ManyArray` currently called a `CollectionRelationship` and soon will just be a `MutableCollection`, and these *are* mutable and their mutations are sent as mutations to the cache. So that is what this test will need to do: test a hasMany relationship. It doesn't hugely matter whether we test `async` or `sync` hasMany relationships, though if testing `async` I would test both and then you'll have to write a `deprecatedTest` for the `async` version since async promise proxies are deprecated now.

Reply:
Thank you! I'll update to test both `async` and `sync` `hasMany` relationships and will include a `deprecatedTest` for `async`. Your explanation is greatly appreciated!